net: mdns: Create a listener to all available network interfaces

jukkar opened the pull request:
Instead of just listening first network interface in the system for mDNS packets, install a multicast listener to all available network interfaces.

Fixes #18748

jukkar committed:
net: context: Allow binding to different interfaces

Allow user to bind to different network interface. This is
useful if binding a multicast address to a certain network
interface.

Signed-off-by: Jukka Rissanen <jukka.rissanen@nordicsemi.no>

jukkar committed:
net: context: Add function to bound to a network interface

Helper function that marks the net_context to bound to a
network interface.

Signed-off-by: Jukka Rissanen <jukka.rissanen@nordicsemi.no>

jukkar committed:
net: conn: Check also network interface for duplicates

When verifying if there are duplicate connections, check
also network interface.

Signed-off-by: Jukka Rissanen <jukka.rissanen@nordicsemi.no>

jukkar committed:
net: if: Add helper to calculate number of interfaces

Add a helper macro that can be used at runtime to return
the number of network interfaces in the system.

Signed-off-by: Jukka Rissanen <jukka.rissanen@nordicsemi.no>

jukkar committed:
net: mdns: Create a listener to all available network interfaces

Instead of just listening first network interface in the system,
install a multicast listener to all available network interfaces.

Fixes #18748

Signed-off-by: Jukka Rissanen <jukka.rissanen@nordicsemi.no>

jukkar committed:
samples: net: mdns_responder: Add VLAN support

Add virtual LAN support to the sample so that it is easier to
test the multiple mDNS listener feature implemented in previous
commit.

Signed-off-by: Jukka Rissanen <jukka.rissanen@nordicsemi.no>

rlubos commented on subsys/net/lib/dns/mdns_responder.c:
```diff
@@ -38,11 +38,11 @@ LOG_MODULE_REGISTER(net_mdns_responder, CONFIG_MDNS_RESPONDER_LOG_LEVEL);
-static struct net_context *ipv4;
-static struct sockaddr_in local_addr4;
+static struct net_context *ipv4[CONFIG_NET_IF_MAX_IPV4_COUNT];
+static struct sockaddr_in local_addr4[CONFIG_NET_IF_MAX_IPV4_COUNT];
```
Do we need multiple `local_addr4` copies? After `setup_ipv4_addr()` we'll end up with an array of the same addresses, and `net_ipv4_igmp_join()` creates a copy anyway.

jukkar replied:
Indeed, good point, will fix.

rlubos commented on subsys/net/lib/dns/mdns_responder.c:
```diff
@@ -645,55 +646,80 @@ static void setup_ipv4_addr(struct sockaddr_in *local_addr)
+	NET_DBG("Setting mDNS listener to %d interface%s", iface_count,
+		iface_count > 1 ? "s" : "");
+
+	if (iface_count > CONFIG_NET_IF_MAX_IPV4_COUNT ||
+	    iface_count > CONFIG_NET_IF_MAX_IPV6_COUNT) {
+		NET_WARN("You have %d interfaces configured but there "
+			 "are %d network interfaces in the system.",
+			 MAX(CONFIG_NET_IF_MAX_IPV4_COUNT,
+			     CONFIG_NET_IF_MAX_IPV6_COUNT), iface_count);
+	}
```
Why would we need such a warning here? Isn't it already covered in `iface_ipv6/4_init()` (it'd also print a warning for similiar case)?

jukkar replied:
The warning is similar but context is a bit different. I think it makes sense to print a warning here, otherwise it might be difficult to track this issue if mDNS does not start properly.

rlubos commented on subsys/net/lib/dns/mdns_responder.c:
```diff
@@ -645,55 +646,80 @@ static void setup_ipv4_addr(struct sockaddr_in *local_addr)
+	setup_ipv6_addr(&local_addr6);
 
-		ipv6 = get_ctx(AF_INET6);
+	for (i = 0; i < CONFIG_NET_IF_MAX_IPV6_COUNT; i++) {
```
Maybe not for this PR, but I wonder wheteher it'd make senese to give applications more control on how mDNS is configured, for example introduce a flag indicating whether mDNS resolver should be configured on an interace or not. But it probably doesn't make sense to implement until really requested (maybe it's not an issues after all to start it unconditionally).

jukkar replied:
Yes, I was thinking the same thing. It is a bit difficult to configure, at least in built time.

cfriedt commented on include/zephyr/net/net_context.h:
```diff
@@ -672,6 +672,23 @@ static inline void net_context_set_iface(struct net_context *context,
 	context->iface = net_if_get_by_iface(iface);
 }
 
+/**
```
Ah, this is cool 👍We've needed this for a while

cfriedt reviewed (dismissed):
Awesome ♥️ You're making me want to port Zephyr to my WRT54GL now 😅

jukkar commented:
Sorry for the reviews, I had to resubmit as one twister check was failing in CI although it was passing locally.

jukkar committed:
net: mdns: Fix compile error when using clang

No issues with gcc but clang gives this error for
the *v4 variable few lines below.

.../lib/dns/mdns_responder.c:712:2: error: expected expression
        struct net_context *v4;

Signed-off-by: Jukka Rissanen <jukka.rissanen@nordicsemi.no>
